node/cn/api.go

```diff
@@ -452,3 +433,86 @@ func (api *PrivateDebugAPI) getModifiedAccounts(startBlock, endBlock *types.Bloc
 	}
 	return dirty, nil
 }
+
+// getStartAndEndBlock returns start and end block based on the given startNum and endNum.
+func (api *PrivateDebugAPI) getStartAndEndBlock(startNum uint64, endNum *uint64) (*types.Block, *types.Block, error) {
+	var startBlock, endBlock *types.Block
+
+	startBlock = api.cn.blockchain.GetBlockByNumber(startNum)
+	if startBlock == nil {
+		return nil, nil, fmt.Errorf("start block number %d not found", startNum)
+	}
+
+	if endNum == nil {
+		endBlock = startBlock
+		startBlock = api.cn.blockchain.GetBlockByHash(startBlock.ParentHash())
+		if startBlock == nil {
+			return nil, nil, fmt.Errorf("block number %d has no parent", startNum)
+		}
+	} else {
+		endBlock = api.cn.blockchain.GetBlockByNumber(*endNum)
+		if endBlock == nil {
+			return nil, nil, fmt.Errorf("end block number %d not found", *endNum)
+		}
+	}
+
+	if startBlock.Number().Uint64() >= endBlock.Number().Uint64() {
+		return nil, nil, fmt.Errorf("start block height (%d) must be less than end block height (%d)", startBlock.Number().Uint64(), endBlock.Number().Uint64())
+	}
+
+	return startBlock, endBlock, nil
+}
+
+// GetModifiedStorageNodesByNumber returns all storage nodes of a contract account
+// that have been changed between the two blocks specified.
+//
+// With one parameter, returns the list of accounts modified in the specified block.
+func (api *PrivateDebugAPI) GetModifiedStorageNodesByNumber(contractAddr common.Address, startNum uint64, endNum *uint64, returnDiffList *bool) ([]common.Address, error) {
+	startBlock, endBlock, err := api.getStartAndEndBlock(startNum, endNum)
+	if err != nil {
+		return nil, err
+	}
+	return api.getModifiedStorageNodes(contractAddr, startBlock, endBlock, returnDiffList)
+}
+
+func (api *PrivateDebugAPI) getModifiedStorageNodes(contractAddr common.Address, startBlock, endBlock *types.Block, returnDiffList *bool) ([]common.Address, error) {
+	startBlockRoot, err := api.cn.blockchain.GetContractStorageRoot(startBlock, api.cn.blockchain.StateCache(), contractAddr)
+	if err != nil {
+		return nil, err
+	}
+	endBlockRoot, err := api.cn.blockchain.GetContractStorageRoot(endBlock, api.cn.blockchain.StateCache(), contractAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	trieDB := api.cn.blockchain.StateCache().TrieDB()
+	oldTrie, err := statedb.NewSecureTrie(startBlockRoot, trieDB)
+	if err != nil {
+		return nil, err
+	}
+	newTrie, err := statedb.NewSecureTrie(endBlockRoot, trieDB)
+	if err != nil {
+		return nil, err
+	}
+
+	diff, _ := statedb.NewDifferenceIterator(oldTrie.NodeIterator([]byte{}), newTrie.NodeIterator([]byte{}))
+	iter := statedb.NewIterator(diff)
+
+	var dirty []common.Address
+	for iter.Next() {
+		key := newTrie.GetKey(iter.Key)
+		if key == nil {
+			return nil, fmt.Errorf("no preimage found for hash %x", iter.Key)
+		}
+		dirty = append(dirty, common.BytesToAddress(key))
+	}
+
+	for _, d := range dirty {
+		logger.Info("modified storage trie nodes", "nodeHash", d.Hash().String())
```

Comment:
Why don't you add more logs printing the number of modified nodes?

Reply:
I think printing all modified nodes in the log seems too verbose.

Reply:
@aidan-kwon Hmm, then how about printing all the modified nodes only if the flag is given? If not, only the number of modified nodes is printed to both console and log.